File: src/main/java/org/radarcns/hdfs/util/Timer.java
package org.radarcns.hdfs.util;

import javax.annotation.Nonnull;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.atomic.LongAdder;
import java.util.stream.Collectors;

public class Timer {
    private final ConcurrentMap<Category, LongAdder> times;
    private static final Timer instance = new Timer();
    private volatile boolean isEnabled;

    public static Timer getInstance() {
        return instance;
    }

    private Timer() {
        this.times = new ConcurrentHashMap<>();
    }

    public void add(String type, long nanoTime) {
        if (isEnabled) {
            Category cat = new Category(type);
            times.computeIfAbsent(cat, c -> new LongAdder()).add(nanoTime);
        }
    }

    public void setEnabled(boolean enabled) {
        isEnabled = enabled;
    }

    @Override
    public String toString() {
        if (!isEnabled) {
            return "Timings: disabled";
        }
        StringBuilder builder = new StringBuilder();
        builder.append("Timings:");

        ConcurrentMap<String, List<Map.Entry<Category, LongAdder>>> timesByType = this.times
                .entrySet().stream()
                .collect(Collectors.groupingByConcurrent(c -> c.getKey().type));

        for (List<Map.Entry<Category, LongAdder>> value : timesByType.values()) {
            long time = value.stream().mapToLong(e -> e.getValue().sum()).sum();
            builder.append("\n\t");
            builder.append(value.get(0).getKey().type);
            builder.append(" - time: ");
            int seconds = (int)(time / 1_000_000_000L);
            int millis = (int) (time / 1_000_000L);
            ProgressBar.formatTime(builder, seconds);
            builder.append('.');
            if (millis < 100) builder.append('0');
            if (millis < 10) builder.append('0');
            builder.append(millis);
            builder.append(" - threads: ");
            builder.append(value.size());
        }

        return builder.toString();
    }

    private static class Category {
        private final String type;
        private final String thread;

        Category(@Nonnull String type) {
            this.type = type;
            this.thread = Thread.currentThread().getName();
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Category category = (Category) o;
            return Objects.equals(type, category.type) &&
                    Objects.equals(thread, category.thread);
        }

        @Override
        public int hashCode() {
            return Objects.hash(type, thread);
        }
    }
}

File: src/main/java/org/radarbase/hdfs/util/Timer.kt
/*
 * Copyright 2018 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarbase.hdfs.util

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap
import java.util.concurrent.atomic.LongAdder

/** Timer for multi-threaded timings. The timer may be disabled to increase program performance.  */
object Timer {
    private val times: ConcurrentMap<String, TimerEntry> = ConcurrentHashMap()

    /**
     * Whether the timer is enabled. A disabled timer will have much less performance impact on
     * timed code.
     */
    @Volatile
    var isEnabled: Boolean = true

    fun <T> time(type: String, action: () -> T): T {
        return if (isEnabled) {
            val startTime = System.nanoTime()
            try {
                action()
            } finally {
                val time = System.nanoTime() - startTime
                times.computeIfAbsent(type) { TimerEntry() }.add(time)
            }
        } else {
            action()
        }
    }

    override fun toString(): String {
        if (!isEnabled) {
            return "Timings: disabled"
        }
        val builder = StringBuilder(100 * times.size)
        builder.append("Timings:")

        this.times.entries
                .forEach { entry ->
                    builder.append("\n\t")
                    builder.append(entry.key)
                    builder.append(" - time: ")
                    formatTime(builder, entry.value.totalTime.sum())
                    builder.append(" - threads: ")
                    builder.append(entry.value.threads.size)
                    builder.append(" - invocations: ")
                    builder.append(entry.value.invocations)
                }

        return builder.toString()
    }

    private class TimerEntry {
        val invocations = LongAdder()
        val totalTime = LongAdder()
        val threads = ConcurrentHashMap<Long, Long>()

        fun add(nanoTime: Long) {
            invocations.increment()
            totalTime.add(nanoTime)
            Thread.currentThread().id.let {
                threads[it] = it
            }
        }
    }

    private fun formatTime(builder: StringBuilder, nanoTime: Long) {
        val seconds = (nanoTime / 1_000_000_000L).toInt()
        val millis = (nanoTime / 1_000_000L).toInt()
        ProgressBar.formatTime(builder, seconds.toLong())
        builder.append('.')
        if (millis < 100) builder.append('0')
        if (millis < 10) builder.append('0')
        builder.append(millis)
    }
}
